pkg/controller/replication/replication_controller.go

```diff
@@ -456,6 +507,50 @@ func (rm *ReplicationManager) worker() {
 	}
 }
 
+func (rm *ReplicationManager) removeControllerRef(task abandon) error {
+	// We need to get the latest pod from the API server because rc manager may
+	// have taken control of the pod, but the update hasn't been observed
+	// through watch yet. If we ignore the pod, the pod will stuck with the
+	// stale controller ref.
+	// TODO: ControlActivePods is running in a different thread, so it's still
+	// possible to falsely take control of a pod.
+	pod, err := rm.kubeClient.Core().Pods(task.podNamespace).Get(task.podName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// ignore the task if the pod doesn't exist anymore.
+			return nil
+		}
+		return err
+	}
+	for _, owner := range pod.OwnerReferences {
+		if owner.UID == task.rcUID && owner.Controller != nil && *owner.Controller {
```

Comment:
Probably I can remove the abandonQueue and do all the "abandon" operation in `syncReplicationController`. If any "abandon" operation fails, I'll add the RC to the workqueue again. This will keep all operations in `syncReplicationController`.

For the defense of the abandonQueue, it's not making the manager stateful, because `syncReplicationController` will go over all the pods and find out what pods need to be abandoned. In other words, if the manager crashes, `syncReplicationController` will drive things back to normal.

The only benefit of keeping the abandon queue is that it makes the update of RC status quicker, because the "abandon" operations are handled in other threads. It seems not enough to justify the added complexity though..

(For the record, at first `syncReplicationController` just lists the pods with matching labels, it cannot find pods with mismatching lables but have stale controllerRef pointing the RC. Hence, I needed the abandonQueue so that I can retry an "abandon" operation if it fails. Now, `syncReplicationController` is able find such pods, abandonQueue is no longer necessary for correctness)